utils/redis.ts

```diff
@@ -0,0 +1,96 @@
+/* eslint-disable no-invalid-this */
+import { identity, pipe } from "fp-ts/lib/function";
+import * as O from "fp-ts/lib/Option";
+import * as redis from "redis";
+import { IConfig } from "./config";
+
+const DEFAULT_REDIS_PORT = "6379";
+
+export type RedisClient = redis.RedisClientType | redis.RedisClusterType;
+
+export class RedisClientFactory {
+  protected readonly config: IConfig;
+  // eslint-disable-next-line functional/prefer-readonly-type
+  protected redisClient: RedisClient | undefined;
+
+  constructor(config: IConfig) {
+    this.config = config;
+  }
+
+  public readonly getInstance = async (): Promise<RedisClient> => {
+    if (!this.redisClient) {
+      // eslint-disable-next-line functional/immutable-data
+      this.redisClient = await pipe(
+        this.config.isProduction,
+        O.fromPredicate(identity),
+        O.chainNullableK(_ => this.config.REDIS_CLUSTER_ENABLED),
+        O.chain(O.fromPredicate(identity)),
+        O.map(() =>
+          this.createClusterRedisClient(
+            this.config.REDIS_URL,
+            this.config.REDIS_PASSWORD,
+            this.config.REDIS_PORT
+          )
+        ),
+        O.getOrElse(() =>
+          this.createSimpleRedisClient(
+            this.config.REDIS_URL,
+            this.config.REDIS_PASSWORD,
+            this.config.REDIS_PORT,
+            this.config.REDIS_TLS_ENABLED
+          )
+        )
+      );
+    }
+    return this.redisClient;
+  };
+
+  protected readonly createSimpleRedisClient = async (
+    redisUrl: string,
+    password?: string,
+    port?: string,
+    useTls: boolean = true
+  ): Promise<RedisClient> => {
+    const redisPort: number = parseInt(port || DEFAULT_REDIS_PORT, 10);
+    const redisClientConnection = redis.createClient<
+      redis.RedisDefaultModules,
+      Record<string, never>,
+      Record<string, never>
+    >({
+      password,
+      socket: {
+        port: redisPort,
+        tls: useTls
+      },
+      url: `redis://${redisUrl}`
+    });
+    await redisClientConnection.connect();
+    return redisClientConnection;
+  };
+
+  protected readonly createClusterRedisClient = async (
+    redisUrl: string,
+    password?: string,
+    port?: string
+  ): Promise<RedisClient> => {
+    const redisPort: number = parseInt(port || DEFAULT_REDIS_PORT, 10);
+    const redisClientConnection = redis.createCluster<
+      redis.RedisDefaultModules,
+      Record<string, never>,
+      Record<string, never>
+    >({
+      defaults: {
+        legacyMode: true,
```

Comment:
> `legacyMode`

what it does and why we are enabling it?

Reply:
This code is simply ported from another repo.
Imho, we should extract our Redis wrapper into a common repo and use it across domains...
There is a long time bad practice to take redis related files and bring them inside repos when we need them, and this is slowing down migrations to newer redis versions, and of course maintenance and new features development.
This time we just kept this bad habit because we are out of time to create a common package.
But I'd track the issue and do it, at least inside our domain to avoid doing this again and again.

Reply:
ok

![image](https://github.com/pagopa/io-functions-service-messages/assets/4357400/1f38b23c-a995-4782-96fc-cd0a27c05690)